c-dependencies/js-compute-runtime/builtins/object-store.cpp

```diff
@@ -0,0 +1,464 @@
+  // because ObjectStore Keys have to be less than 1025 characters.
+  if (key_len > 1024) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr, JSMSG_OBJECT_STORE_KEY_TOO_LONG);
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  char *key_chars = key.get();
+
+  if (strchr(key_chars, '#') != NULL) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                              JSMSG_OBJECT_STORE_KEY_INVALID_CHARACTER, "#");
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+  if (strchr(key_chars, '?') != NULL) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                              JSMSG_OBJECT_STORE_KEY_INVALID_CHARACTER, "?");
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+  if (strchr(key_chars, '*') != NULL) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                              JSMSG_OBJECT_STORE_KEY_INVALID_CHARACTER, "*");
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+  if (strchr(key_chars, '[') != NULL) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                              JSMSG_OBJECT_STORE_KEY_INVALID_CHARACTER, "[");
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+  if (strchr(key_chars, ']') != NULL) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                              JSMSG_OBJECT_STORE_KEY_INVALID_CHARACTER, "]");
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+  if (strchr(key_chars, '\n') != NULL) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                              JSMSG_OBJECT_STORE_KEY_INVALID_CHARACTER, "newline");
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+  if (strchr(key_chars, '\r') != NULL) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                              JSMSG_OBJECT_STORE_KEY_INVALID_CHARACTER, "carriage return");
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+  auto acme_challenge = ".well-known/acme-challenge/";
+  if (strncmp(key_chars, acme_challenge, strlen(acme_challenge)) == 0) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr, JSMSG_OBJECT_STORE_KEY_ACME);
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  if (strcmp(key_chars, ".") == 0 || strcmp(key_chars, "..") == 0) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr, JSMSG_OBJECT_STORE_KEY_RELATIVE);
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  BodyHandle body_handle = {INVALID_HANDLE};
+  int status =
+      fastly_object_store_lookup(object_store_handle(self), key_chars, key_len, &body_handle);
+  if (!HANDLE_RESULT(cx, status)) {
+    return false;
+  }
+
+  // If the handle is invalid it means no entry was found
+  // When no entry is found, we are going to resolve the Promise with `null`.
```

Comment:
Should we reject instead?

Reply:
From reading the advice of the W3C TAG (Technical Architecture Group) document on [Writing Promise-Using Specifications](https://www.w3.org/2001/tag/doc/promises-guide#rejections-should-be-exceptional) I thought we should resolve the promise in this situation as the I/O has succeeded but the value was not found, which is a specific example they have about when to resolve and not reject